File: adventofcode/year2016/day23.py
from __future__ import annotations
from adventofcode.common import Solution
from copy import copy
from typing import List


class OptimizedCPU(object):
    def __init__(self, instructions: List[str]):
        self._instructions = copy(instructions)
        self._registers = {'a': 0,  'b': 0, 'c': 0, 'd': 0}
        self._verbose = False

    @property
    def a(self):
        return self._registers['a']

    @a.setter
    def a(self, value: int):
        self._registers['a'] = value

    @property
    def b(self):
        return self._registers['b']

    @b.setter
    def b(self, value: int):
        self._registers['b'] = value

    @property
    def c(self):
        return self._registers['c']

    @c.setter
    def c(self, value: int):
        self._registers['c'] = value

    @property
    def d(self):
        return self._registers['d']

    @d.setter
    def d(self, value: int):
        self._registers['d'] = value

    @property
    def instructions(self):
        return self._instructions

    def verbose(self, toggle: bool) -> None:
        self._verbose = toggle

    def run(self) -> None:
        index = 0
        while index < len(self._instructions):

            # look ahead to optimize incremental loops
            if index + 4 < len(self._instructions):
                match [self._instructions[index + i].split(' ') for i in range(5)]:
                    case ['inc', x1], ['dec', x2], ['jnz', _, '-2'], ['dec', _], ['jnz', x5, '-5']:
                        # multiplication sequence can be optimized to x1 += x2 * x5
                        self._registers[x1] += self._registers[x2] * self._registers[x5]

                        if self._verbose:
                            print(f"a:{self.a} b:{self.b} c:{self.c} d:{self.d} @ [{index}] optimized 5 instructions to {x1} += {x2} * {x5}")

                        index += 5
                        continue
                    case ['dec', x1], ['inc', x2], ['jnz', _, '-2'], _, _:
                        # copy sequence can be optimized to x2 += x1
                        self._registers[x2] += self._registers[x1]

                        if self._verbose:
                            print(f"a:{self.a} b:{self.b} c:{self.c} d:{self.d} @ [{index}] optimized 3 instructions to {x2} += {x1}")

                        index += 3
                        continue
                    case _:
                        pass

            match self._instructions[index].split(' '):
                case 'cpy', x, y:
                    if y in self._registers:
                        self._registers[y] = self._registers[x] if x in self._registers else int(x)
                    if self._verbose:
                        print(f"a:{self.a} b:{self.b} c:{self.c} d:{self.d} @ [{index}] {self._instructions[index]}")
                    index += 1
                case 'inc', x:
                    if x in self._registers:
                        self._registers[x] += 1
                    if self._verbose:
                        print(f"a:{self.a} b:{self.b} c:{self.c} d:{self.d} @ [{index}] {self._instructions[index]}")
                    index += 1
                case 'dec', x:
                    if x in self._registers:
                        self._registers[x] -= 1
                    if self._verbose:
                        print(f"a:{self.a} b:{self.b} c:{self.c} d:{self.d} @ [{index}] {self._instructions[index]}")
                    index += 1
                case 'jnz', x, y:
                    x = self._registers[x] if x in self._registers else int(x)
                    y = self._registers[y] if y in self._registers else int(y)
                    if self._verbose:
                        print(f"a:{self.a} b:{self.b} c:{self.c} d:{self.d} @ [{index}] {self._instructions[index]}")
                    index += y if x != 0 else 1
                case 'tgl', x:
                    x = self._registers[x] if x in self._registers else int(x)
                    if 0 <= (index + x) < len(self._instructions):
                        match self._instructions[index + x].split(' '):
                            case 'cpy', _, _:
                                self._instructions[index + x] = self._instructions[index + x].replace('cpy', 'jnz')
                            case 'inc', _:
                                self._instructions[index + x] = self._instructions[index + x].replace('inc', 'dec')
                            case 'dec', _:
                                self._instructions[index + x] = self._instructions[index + x].replace('dec', 'inc')
                            case 'jnz', _, _:
                                self._instructions[index + x] = self._instructions[index + x].replace('jnz', 'cpy')
                            case 'tgl', _:
                                self._instructions[index + x] = self._instructions[index + x].replace('tgl', 'inc')
                            case _:
                                raise Exception(f"Unrecognized instruction : {self._instructions[index + x]}")
                    if self._verbose:
                        print(f"a:{self.a} b:{self.b} c:{self.c} d:{self.d} @ [{index}] {self._instructions[index]}")
                    index += 1
                case _:
                    raise Exception(f"Unrecognized instruction : {self._instructions[index]}")


class Day23(Solution):
    def __init__(self, year: str, day: str):
        super().__init__(year, day)
        self._instructions = self._load_input_as_lines()

    def part_one(self):
        cpu = OptimizedCPU(self._instructions)
        cpu.verbose(True)
        cpu.a = 7
        cpu.run()

        return cpu.a

    def part_two(self):
        cpu = OptimizedCPU(self._instructions)
        cpu.verbose(True)
        cpu.a = 12
        cpu.run()

        return cpu.a

File: adventofcode/year2017/Day23.cs
using adventofcode.common;

namespace adventofcode.year2017;

public class Day23 : Solution
{
    private string[] _input;
    
    public Day23(string year, string day) : base(year, day)
    {
        _input = LoadInputAsLines();
    }

    public override string PartOne()
    {
        var cp = new Coprocessor(_input);
        cp.Run(false);
        return Convert.ToString(cp.InstructionCount("mul"));
    }

    public override string PartTwo()
    {
        var cp = new Coprocessor(_input);
        cp.SetRegister("a", 1L);
        cp.Optimize(true);
        cp.Run(true);
        return Convert.ToString(cp.Register("h"));
    }
    
    private class Coprocessor
    {
        private List<string> _instructions;
        private Dictionary<string, long> _registers;
        private Dictionary<string, long> _instructionCount;
        private bool _optimize;

        public Coprocessor(string[] instructions)
        {
            _optimize = false;
            _instructions = new List<string>(instructions);
            _instructionCount = new Dictionary<string, long>()
            {
                {"set", 0}, {"sub", 0}, {"mul", 0}, {"jnz", 0}
            };
            _registers = new Dictionary<string, long>()
            {
                {"a", 0}, {"b", 0}, {"c", 0}, {"d", 0}, {"e", 0}, {"f", 0}, {"g", 0}, {"h", 0}
            };
        }

        private long Value(string value)
        {
            return _registers.ContainsKey(value) ? _registers[value] : Convert.ToInt32(value);
        }

        public void Optimize(bool optimize) { _optimize = optimize; }

        public long Register(string register) { return _registers[register]; }

        public void SetRegister(string register, long value) { _registers[register] = value; }

        public long InstructionCount(string instruction) { return _instructionCount[instruction]; }

        public void Run(bool verbose)
        {
            var index = 0L;
            while (index >= 0L && index < _instructions.Count)
            {
                if (_optimize)
                { 
                    //optimization
                    var lookAhead = Enumerable.Range((int) index, (int) Math.Min(_instructions.Count - index, 16)).Select(i => _instructions[i]).ToList();
                    switch (lookAhead)
                    {
                        case ["set f 1", "set d 2", "set e 2", "set g d", "mul g e", "sub g b", "jnz g 2", "set f 0", "sub e -1", "set g e", "sub g b", "jnz g -8", "sub d -1", "set g d", "sub g b", "jnz g -13"]:
                            // instruction sequence sets f to 0 if b = d * e, where d and e are greater than 2. otherwise, set f to 1
                            _registers["f"] = 1;
                            for (var x = 2; x < _registers["b"] / 2; x++)
                            {
                                if (_registers["b"] % x == 0)
                                {
                                    _registers["f"] = 0;
                                    break;
                                }
                            }
                            _registers["d"] = _registers["b"];
                            _registers["e"] = _registers["b"];
                            _registers["g"] = 0;
                            if (verbose) { Console.WriteLine($"{index + 1}: [{_instructions[(int) index]}] {string.Join(", ", _registers.Select(kv => $"{kv.Key}: {kv.Value}"))} optimized 16 instructions"); }
                            index += 16;
                            break;
                    }
                }

                var parts = _instructions[(int) index].Split(" ");
                switch (parts[0])
                {
                    case "set":
                        _registers[parts[1]] = Value(parts[2]);
                        if (verbose) { Console.WriteLine($"{index + 1}: [{_instructions[(int) index]}] {string.Join(", ", _registers.Select(kv => $"{kv.Key}: {kv.Value}"))}"); }
                        _instructionCount["set"]++;
                        break;
                    case "sub":
                        _registers[parts[1]] -= Value(parts[2]);
                        if (verbose) { Console.WriteLine($"{index + 1}: [{_instructions[(int) index]}] {string.Join(", ", _registers.Select(kv => $"{kv.Key}: {kv.Value}"))}"); }
                        _instructionCount["sub"]++;
                        break;
                    case "mul":
                        _registers[parts[1]] *= Value(parts[2]);
                        if (verbose) { Console.WriteLine($"{index + 1}: [{_instructions[(int) index]}] {string.Join(", ", _registers.Select(kv => $"{kv.Key}: {kv.Value}"))}"); }
                        _instructionCount["mul"]++;
                        break;
                    case "jnz":
                        if (verbose) { Console.WriteLine($"{index + 1}: [{_instructions[(int) index]}] {string.Join(", ", _registers.Select(kv => $"{kv.Key}: {kv.Value}"))}"); }
                        index += Value(parts[1]) != 0 ? Value(parts[2]) : 1;
                        _instructionCount["jnz"]++;
                        continue;
                    default:
                        throw new Exception($"Invalid command : {parts[0]}");
                }
                index += 1;
            }
        }
    }
}